File: 08-palindromic-squares/palsquare.cpp
/*
ID: caleb.h1
LANG: C++
PROG: palsquare
*/

#include <string>
#include <fstream>
#include <iostream>

// Map numbers to higher base digits
// 0 -> 0, 5 -> 5
// 10 -> a, 15 -> f
char convertNumToDigit(char c) {
    if (c < 10)
        return c + '0';
    return c - 10 + 'A';
}

std::string tenToBase(unsigned int number, char base) {
    std::string newNumber;
    
    while (number > 0) {
        newNumber = convertNumToDigit(number % base) + newNumber;
        number /= base;
    }

    return newNumber;
}


bool isPalindrome(std::string s) {
    for (int i = 0; i < s.length() / 2; ++i) {
        if (s[i] != s[s.length()-i-1])
            return false;
    }
    return true;
}


int main(int argc, const char* argv[]) {

    std::ifstream infile("palsquare.in");
    int base;
    infile >> base;
    infile.close();

    std::ofstream outfile("palsquare.out");
    for (int i = 1; i < 301; ++i) {
        int square = i * i;
        std::string squareString = tenToBase(square, base);
        // std::cout << i << ' ' << square << ' ' << squareString << std::endl;
        if (isPalindrome(squareString))
            outfile << tenToBase(i, base) << ' ' << squareString << std::endl;
    }

    outfile.close();

    return 0;
}

File: 08-palindromic-squares/palsquare.py
"""
ID: caleb.h1
LANG: PYTHON3
PROG: palsquare
"""

# Map numbers to higher base digits
# 0 -> 0, 5 -> 5
# 10 -> a, 15 -> f
def convertNumToDigit(c: int) -> str:
    if c < 10:
        return str(c)
    return chr(c - 10 + ord('A'))

def tenToBase(number: int, base: int) -> str:
    newNumber: str = ""
    
    while (number > 0):
        newNumber = convertNumToDigit(number % base) + newNumber
        number //= base

    return newNumber

def isPalindrome(s: str) -> bool:
    for i in range(len(s) // 2):
        if s[i] != s[len(s) - i - 1]:
            return False
    
    return True


def main() -> None:

    with open("palsquare.in", "r") as f:
        base: int = int(f.readline())

    with open("palsquare.out", "w") as f:
        for i in range(1, 301):
            square: int = i * i
            squareString: str = tenToBase(square, base)
            if isPalindrome(squareString):
                f.write(f'{tenToBase(i, base)} {squareString}\n')

if __name__ == '__main__':
    main()

